core/src/cas/eav.rs

```diff
@@ -0,0 +1,173 @@
+use error::HolochainError;
+use cas::content::Address;
+use std::collections::HashSet;
+
+/// EAV (entity-attribute-value) data
+/// ostensibly for metadata about entries in the DHT
+/// defines relationships between AddressableContent values
+/// implemented on top of cas::storage::ContentAddressableStorage
+/// @see https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model
+/// Address of AddressableContent representing the EAV entity
+type Entity = Address;
+
+/// using String for EAV attributes (not e.g. an enum) keeps it simple and open
+type Attribute = String;
+
+/// Address of AddressableContent representing the EAV value
+type Value = Address;
+
+// @TODO do we need this?
+// unique (local to the source) monotonically increasing number that can be used for crdt/ordering
+// @see https://papers.radixdlt.com/tempo/#logical-clocks
+// type Index ...
+
+// @TODO do we need this?
+// source agent asserting the meta
+// type Source ...
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct EntityAttributeValue {
+    entity: Entity,
+    attribute: Attribute,
+    value: Value,
+    // index: Index,
+    // source: Source,
+}
```

Comment:
I get and like the idea behind this struct. What I find questionnable is the name and how it is presented. Since value is constrained to an Address and the whole thing is for `AddressableContent`, I think it would clearer to have that reflected in the name. The current name `EntityAttributeValue` implies a general purpose struct but it is not, since its specific.

I would suggest something like `AddressableContentLink`, and have the same field names as `Link` (base, target, tag).
Maybe use `Relation` or `Edge` instead of `Link` to not confuse the two?

Reply:
I like `AddressableContentRelationship`.  You can still leave all the EAV references in the comments.

Reply:
@ddd-mtl looks like EAV is also being used in the network proposal hackmd from @lucksus

Reply:
Hm, I would actually prefer sticking to known vocabulary. And yes, with these traits we are describing the general case, which we then implement for `Links` or `Relations`.

```rust
impl EntityAttributValue for DHT::EntryLink {
...
}
```
no?

Reply:
`EntityAttributValue` is a struct not a trait so your example would't compile ^^'''
I'm not sure if you are mistaken because you might be tired or because the name is indeed misleading and you just confirmed it.

Anyhow if you want to be descriptive to known vocabulary, this is not EAV since V **must** be an address like E, so its more like EAE. In a graph database thats a directed and named Edge / Relationship.

Reply:
@lucksus yes, it's the general case that we can implement for `Links` or `Relations` _and_ `Address` is general enough to accept any value that can implement `AddressableContent`.

but, `EntityAttributeValue` is a struct not a trait!

`attribute` defines the "type". It is currently a `String` (e.g. `"link"`) to keep attributes open, but we could use an enum if we want the allowed values to be closed (e.g. `Attribute::Link`).

here's a contrived example implementing a relationship between `"1"` and `"2"` to show that 1 comes before 2.

```rust
impl AddressableContent for String {
  fn content (&self) -> Content { self.clone() }
  fn from_content (content: &Content) -> Self { content.clone() }
}

let store = ExampleEntityAttributeValueStorage::new();
let 1_before_2 = EntityAttributeValue::new("1".to_string().address(), "comes_before".to_string(), "2".to_string().address());
store.add(1_before_2);
```

Reply:
Thanks for explanation. I finally get it 👍